tools/pscript/cli.c

```diff
@@ -0,0 +1,276 @@
+ **/
+static int _scan_brackets(pss_comp_lex_t* lexer)
+{
+	static pss_comp_lex_token_type_t bracket_stack[MAX_BRACKETS];
+	static int b_index;
+	pss_comp_lex_token_t token;
+	while(1)
+	{
+		if(ERROR_CODE(int) == pss_comp_lex_next_token(lexer, &token))
+		{
+			ERROR_RETURN_ACTION(int, b_index = 0);
+		}
+		switch(token.type)
+		{
+			case PSS_COMP_LEX_TOKEN_LPARENTHESIS:
+			case PSS_COMP_LEX_TOKEN_LBRACKET:
+			case PSS_COMP_LEX_TOKEN_LBRACE:
+			    bracket_stack[b_index++] = token.type;
+			    break;
+			case PSS_COMP_LEX_TOKEN_RPARENTHESIS:
+			    _CHECK_BRACKETS_TOP(PSS_COMP_LEX_TOKEN_LPARENTHESIS)
+			case PSS_COMP_LEX_TOKEN_RBRACKET:
+			    _CHECK_BRACKETS_TOP(PSS_COMP_LEX_TOKEN_LBRACKET)
+			case PSS_COMP_LEX_TOKEN_RBRACE:
+			    _CHECK_BRACKETS_TOP(PSS_COMP_LEX_TOKEN_LBRACE)
+			case PSS_COMP_LEX_TOKEN_ERROR:
+			    ERROR_RETURN_ACTION(int, b_index = 0);
+			case PSS_COMP_LEX_TOKEN_EOF:
+			    return b_index;
+			default:
+			    /* Ignore */
+			    break;
+		}
+		// check stack size
+		if(b_index >= MAX_BRACKETS)
+		{
+			LOG_ERROR("Code too long");
+			ERROR_RETURN_ACTION(int, b_index = 0);
+		}
+	}
+	return b_index;
+}
+
+/** @brief add one line to the line_list
+ * @return the new head
+ */
+static struct _line_list_t* _add_line(struct _line_list_t* head, char* line, uint32_t size, uint32_t off)
+{
+	if(NULL == line)
+	    return head;
+
+	struct _line_list_t *node = (struct _line_list_t*)malloc(sizeof(*node));
+	if(NULL == node)
+	    return NULL;
+	node->line = line;
+	node->size = size;
+	node->off = off;
+	node->next = head;
+	return node;
+}
+
+static void _stop(int signo)
+{
+	(void)signo;
+	LOG_DEBUG("SIGINT Caught!");
+	_interrupt = 1;
+}
+
+int pss_cli_interactive(uint32_t debug)
+{
+	static const char* source_path = "_";
+	char* line = (char*)NULL;
+	pss_comp_lex_t* lexer = NULL;
+	pss_bytecode_module_t* module = NULL;
+	pss_comp_error_t* err = NULL;
+	struct _line_list_t *head = NULL;
+	char *code;
+	uint32_t lex_success;
+
+	current_vm = pss_vm_new();
+	if(NULL == current_vm || ERROR_CODE(int) == builtin_init(current_vm))
+	{
+		if(current_vm != NULL) pss_vm_free(current_vm);
+		LOG_ERROR("Cannot create PSS Virtual Machine");
+		return 1;
+	}
+	uint32_t code_size;
+
+	signal(SIGINT, _stop);
+
+	while(1)
+	{
+		line = readline(_prompt);
+		// ignore empty line
+		if(NULL == line)
+		    continue;
+		if(0 == *line)
+		{
+			free(line);
+			continue;
+		}
+
+		// quit TODO: builtin commands
+		if('q' == *line && 0 == *(line + 1))
+		{
+			free(line);
+			return 0;
+		}
+
+		code_size = 0;
+		head = NULL;
+		module = NULL;
+		lexer = NULL;
+		lex_success = 0;
+		err = NULL;
+
+		while(NULL != line && 0 == _interrupt)
+		{
+			uint32_t line_size = (uint32_t)strlen(line);
+			head = _add_line(head, line, line_size, code_size);
+			if(NULL == head)
+			{
+				LOG_ERROR("malloc _line_list_t failed");
+				break;
+			}
+			// should add a newline at the end of each line
+			code_size += line_size + 1;
```

Comment:
I would prefer not having a code_size variable, instead, change the semantics of line_list->size from "the length of the this line exclude the terminating char" to "the length including the terminating char"